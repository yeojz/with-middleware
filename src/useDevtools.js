import createDevtoolsListener from './helpers/createDevtoolsListener';

let instanceId = 0;
const instances = {};
const initialized = {};

function getInstanceId(title) {
  if (title) {
    return title;
  }

  instanceId += 1;
  return instanceId;
}

function getInstance(instanceId) {
  if (instances[instanceId]) {
    return instances[instanceId];
  }

  if (typeof window !== 'undefined' && window.__REDUX_DEVTOOLS_EXTENSION__) {
    instances[instanceId] = window.__REDUX_DEVTOOLS_EXTENSION__.connect({
      name: instanceId
    });

    return instances[instanceId];
  }
}

function subscribe(store, next, instance, instanceId) {
  if (instance && !initialized[instanceId]) {
    instance.init(store.getState());

    const listener = createDevtoolsListener(store, next, instance);

    instance.subscribe(listener);
    initialized[instanceId] = true;
  }
}

function useDevtools(title) {
  const instanceId = getInstanceId(title);
  const instance = getInstance(instanceId);

  return store => next => action => {
    subscribe(store, next, instance, instanceId);

    const state = next(action);
    if (instance) {
      instance.send(action, store.getState());
    }
    return state;
  };
}

export default useDevtools;
