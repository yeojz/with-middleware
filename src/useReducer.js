import { BATCH_ACTION, SET_ACTION } from './constants';

function batchAction(reducer, store, next, action) {
  const state = action.payload.reduce(
    (nextState, current) => reducer(nextState, current),
    action.state || store.getState()
  );

  return next(state);
}

function useReducer(reducer) {
  return store => next => action => {
    if (action.type === BATCH_ACTION) {
      return batchAction(reducer, store, next, action);
    }

    if (action.type === SET_ACTION) {
      return next(action.payload);
    }

    const state = reducer(store.getState(), action);
    return next(state);
  };
}

export default useReducer;
