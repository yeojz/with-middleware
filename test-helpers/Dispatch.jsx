/* eslint-disable react/prop-types */
import React from 'react';

function Dispatch(props) {
  return (
    <div>
      <div data-testid="value">{props.state.count}</div>
      <button
        data-testid="increment"
        onClick={() => props.dispatch({ type: 'INCREMENT' })}
      >
        Increment
      </button>
      <button
        data-testid="decrement"
        onClick={() => props.dispatch({ type: 'DECREMENT' })}
      >
        Decrement
      </button>
    </div>
  );
}

export default Dispatch;
