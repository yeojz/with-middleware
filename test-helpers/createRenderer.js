import React from 'react';
import { compose, withState } from 'recompose';
import { render } from 'react-testing-library';
import withMiddleware from '../src/withMiddleware';

function createRenderer(BaseComponent) {
  return (stateName, stateUpdaterName, initialState, middlewares = []) => {
    const Component = compose(
      withState(stateName, stateUpdaterName, initialState),
      withMiddleware(stateName, stateUpdaterName, middlewares)
    )(BaseComponent);

    return render(<Component />);
  };
}

export default createRenderer;
