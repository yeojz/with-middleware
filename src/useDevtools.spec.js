import { cleanup, fireEvent } from 'react-testing-library';
import createRenderer from '../test-helpers/createRenderer';
import Dispatch from '../test-helpers/Dispatch';
import reducer from '../test-helpers/reducer';
import useReducer from './useReducer';
import useDevtools from './useDevtools';

jest.mock('./helpers/createDevtoolsListener', () => jest.fn());

describe('useDevtools - mounted', () => {
  const connect = {
    send: jest.fn(),
    init: jest.fn(),
    subscribe: jest.fn()
  };

  beforeEach(() => {
    connect.init.mockReset();
    connect.send.mockReset();
    connect.subscribe.mockReset();
    global.__REDUX_DEVTOOLS_EXTENSION__ = {
      connect: () => connect
    };
  });

  afterEach(cleanup);

  const renderer = createRenderer(Dispatch);

  function createWithDevtool(title) {
    return renderer('state', 'dispatch', reducer(), [
      useDevtools(title),
      useReducer(reducer)
    ]);
  }

  test('should render without errors (with title)', () => {
    const { container } = createWithDevtool('test');
    expect(container.firstChild).toMatchSnapshot();
  });

  test('should render without errors (without title)', () => {
    const { container } = createWithDevtool();
    expect(container.firstChild).toMatchSnapshot();
  });

  test('should send correct information to redux dev tool', () => {
    const { getByTestId } = createWithDevtool('test');
    fireEvent.click(getByTestId('increment'));
  });
});
