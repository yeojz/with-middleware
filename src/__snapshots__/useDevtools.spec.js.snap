// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useDevtools - mounted should render without errors (with title) 1`] = `
<div>
  <div
    data-testid="value"
  >
    0
  </div>
  <button
    data-testid="increment"
  >
    Increment
  </button>
  <button
    data-testid="decrement"
  >
    Decrement
  </button>
</div>
`;

exports[`useDevtools - mounted should render without errors (without title) 1`] = `
<div>
  <div
    data-testid="value"
  >
    0
  </div>
  <button
    data-testid="increment"
  >
    Increment
  </button>
  <button
    data-testid="decrement"
  >
    Decrement
  </button>
</div>
`;
